core/jvm/src/main/scala/zio/internal/ZScheduler.scala

```diff
@@ -511,6 +481,29 @@ private[zio] object ZScheduler {
     def put(trace: Trace): Long
   }
 
+  private object Locations {
+
+    final class Enabled(initialSize: Int = 64) extends Locations {
+      private[this] val locations = mutable.HashMap.empty[Trace, AtomicLong]
+      locations.sizeHint((initialSize / 0.75d).toInt)
+
+      def get(trace: Trace): Long = {
+        val v = locations.getOrElse(trace, null)
+        if (v eq null) 0L else v.get()
+      }
+
+      def put(trace: Trace): Long =
+        locations
+          .getOrElseUpdate(trace, new AtomicLong(0L))
+          .getAndIncrement()
```

Comment:
Minor: We could keep the previous logic to avoid a call to `.getAndIncrement`:

```scala
def put(trace: Trace): Long = {
        val v = locations.getOrElse(trace, null)
        if (v eq null) {
          locations.put(trace, new AtomicLong(1L))
          0L
        } else v.getAndIncrement()
      }
```

Reply:
The main issue with this implementation is that it's not thread-safe. 2 threads that call `getOrElse` at the same time can both return `null`, which means that both will call `put` at the same time with the last call overriding the entry